src/internal/operators/catchError.ts

```diff
@@ -111,44 +111,29 @@ export function catchError<T, O extends ObservableInput<any>>(
   return (source: Observable<T>) =>
     lift(source, function (this: Subscriber<T>, source: Observable<T>) {
       const subscriber = this;
-      const subscription = new Subscription();
-      let innerSub: Subscription | null = null;
-      let syncUnsub = false;
-      let handledResult: Observable<ObservedValueOf<O>>;
+      let nextSource: Observable<T> | null = null;
+      let subscription: Subscription;
 
-      const handleError = (err: any) => {
-        try {
-          handledResult = from(selector(err, catchError(selector)(source)));
-        } catch (err) {
-          subscriber.error(err);
-          return;
+      const tryToSwapSubscription = () => {
+        if (subscription && nextSource && !subscriber.closed) {
+          subscription = nextSource.subscribe(subscriber);
         }
       };
 
-      innerSub = source.subscribe(
+      subscription = source.subscribe(
         new CatchErrorSubscriber(subscriber, (err) => {
-          handleError(err);
-          if (handledResult) {
-            if (innerSub) {
-              innerSub.unsubscribe();
-              innerSub = null;
-              subscription.add(handledResult.subscribe(subscriber));
-            } else {
-              syncUnsub = true;
-            }
+          try {
+            nextSource = from(selector(err, catchError(selector)(source)));
+            tryToSwapSubscription();
```

Comment:
There's another interesting semantic difference here. Any error that occurs in `tryToSwapSubscription()` will be caught and sent down the `error` channel... I'm not sure what this could mean. In the event that someone is using the deprecated sync error throwing (set in by `config`), it's possible that they have a synchronous error thrown during this subscription... I'm not sure what the behavior in this case here would be.. (And I'm equally unsure what the behavior would be in the previous code).

I guess I would expect it to send down the error channel as you do here, so this is probably correct. It's just something to think about.

Probably nothing we should be concerned with, I suppose. But I wanted to note the concern.